pkg/secrets/secrets.go

```diff
@@ -33,37 +36,56 @@ func DecodeAsFile(ctx context.Context, val string) (string, error) {
 	return f.Name(), nil
 }
 
-type decrypter func(val string) (string, error)
+type decrypter func(c client.Client, namespace string, val string) (string, error)
 
 func decode(decrypt decrypter, ctx context.Context, val string) (string, error) {
 	if !isSecretEncrypted(val) {
 		return val, nil
 	}
 
 	var v string
-	c, cacheOk := FromContext(ctx)
-	if cacheOk {
-		// Check if in cache
-		if v, ok := (*c)[val]; ok {
-			return v, nil
-		}
-	}
-	v, err := decrypt(val)
+	c, err := FromContextWithError(ctx)
 	if err != nil {
 		return "", err
 	}
-	if cacheOk {
-		// If we could get the cache, update it
-		(*c)[val] = v
+	// Check if in cache
+	if v, ok := c.Cache[val]; ok {
+		return v, nil
 	}
+	v, err = decrypt(c.Client, c.Namespace, val)
+	if err != nil {
+		return "", err
+	}
+	// If we could get the cache, update it
+	c.Cache[val] = v
 	return v, nil
 }
 
 func isSecretEncrypted(str string) bool {
 	return strings.HasPrefix(str, "encrypted:")
 }
 
-func decryptFunc(val string) (string, error) {
-	secretDecrypter := secrets.NewDecrypter(val)
+func decryptFunc(c client.Client, namespace string, val string) (string, error) {
+	secretDecrypter, err := NewDecrypter(c, namespace, val)
+	if err != nil {
+		return "", err
+	}
 	return secretDecrypter.Decrypt()
 }
+
+func NewDecrypter(c client.Client, namespace string, encryptedSecret string) (secrets.Decrypter, error) {
+	engine, params := secrets.ParseTokens(encryptedSecret)
+	switch engine {
+	case "s3":
+		return secrets.NewS3Decrypter(params), nil
+	case "vault":
+		return secrets.NewVaultDecrypter(params), nil
+	case "k8s":
+		return kubernetes.NewSecretDecrypter(c, namespace, params), nil
+	case "noop":
```

Comment:
What's a `noop` secret?

Reply:
noop = nop above. No operation decrypted.

Reply:
I'm still not sure how this works in real life or why is needed, is it just to support some test cases? there could be another way to test without creating a dummy secrets engine.

Reply:
I use it for testing and don't intent to publish it. The only use case I'd see is someone inlining the kubeconfig as a string and not wanting to use a Kubernetes secret.